### DISCUSSION TITLE
FileSystemProvider doesn't work with Intellisense

### DISCUSSION THREAD

**Discussion by diggyk** [microsoft/vscode-discussions#385] [Extension Development QnA] [answered]
CONTEXT: My extension [Fildir](https://marketplace.visualstudio.com/items?itemName=diggyk.fildir) provides a custom FileSystemProvider to give a filtered view of files available in from `vscode.workspace.fs`, but it breaks Intellisense and I can't figure out why or how to fix it.

Hello!  I've created a FileSystemProvider extension so people can filter out files/directories they don't want to see.  It basically still uses the file system but responds to `readDirectory` calls by only return the directories and files that match the filters given. But when any other call is made (like `readFile`), it just translates from the `fildir:/` URI to the `file:/` URI. 

But I have a problem! The issue is that Intellisense stops working, I think because it doesn't "see" all the files for a project because they may be filtered out. 

In my `readFile` call, i can do a `vscode.window.showTextDocument(Uri.file(path));` and that works since it's the actual file opened out of the actual workspace root so Intellisense can see everything.  But it creates two problems:

1. I still have to return some data from `readFile` and as such, a text document will always open, even though I don't want it to. 
2. The file explorer scrolls back up to the workspace that the file originated from instead of just staying where it was (e.g. at my "Filtered Directories" workspace root)

What are some options here?  Should I just give up on the FSP and use a separate panel with a Tree and just do my best to mimick a File Explorer?

**Comment by r3m0t** [comment, marked as answer]
You can implement request forwarding? Your document basically embeds another document. Implement a hover provider that requests hover information against the original filename and returns it. You don't even need to adjust any line or column numbers.

https://code.visualstudio.com/api/language-extensions/embedded-languages

**Reply by diggyk** [reply]
If I'm understanding your recommendation, I can try it. But that would mean I use a custom extension for each file, no? So in the file explorer, you wouldn't see the real name and the real icon for the file type, correct?

**Reply by r3m0t** [reply]
No, you aren't changing the extension, you're changing the URI scheme.

Here's an example, you would implement about 30 of these to cover all the features - take a look here - https://code.visualstudio.com/api/references/commands

```typescript

	vscode.languages.registerHoverProvider(
		[{scheme: 'fildir'}],
		{
			provideHover(document, position, token) {
				const {actualUri} = parseFildirUri(document.uri);
				return vscode.commands.executeCommand("vscode.executeHoverProvider", actualUri, position);
			},
		}
	);
	vscode.languages.registerDocumentLinkProvider(
		[{scheme: 'fildir'}],
		{
			provideDocumentLinks(document, token) {
				const {actualUri} = parseFildirUri(document.uri);
				return vscode.commands.executeCommand("vscode.executeLinkProvider", actualUri);

			},
		}
	);
```
